File: app_manager/appmanager/src/main/java/app/editors/manager/ui/adapters/diffutilscallback/EntityDiffUtilsCallback.java
package app.editors.manager.ui.adapters.diffutilscallback;

import java.util.List;

import app.editors.manager.mvp.models.base.Entity;
import app.editors.manager.mvp.models.explorer.File;
import app.editors.manager.mvp.models.explorer.Folder;
import app.editors.manager.mvp.models.list.Footer;
import app.editors.manager.mvp.models.list.Header;

public class EntityDiffUtilsCallback extends BaseDiffUtilsCallback<Entity> {

    public EntityDiffUtilsCallback(List<Entity> mNewList, List<Entity> mOldList) {
        super(mNewList, mOldList);
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        Entity oldEntity = mOldList.get(oldItemPosition);
        Entity newEntity = mNewList.get(newItemPosition);
        if (oldEntity instanceof Header && newEntity instanceof Header) {
            return true;
        } else if (oldEntity instanceof File && newEntity instanceof File) {
            File newFile = (File) newEntity;
            File oldFile = (File) oldEntity;
            return newFile.getId().equals(oldFile.getId());
        } else if (oldEntity instanceof Folder && newEntity instanceof Folder) {
            Folder newFolder = (Folder) newEntity;
            Folder oldFolder = (Folder) oldEntity;
            return newFolder.getId().equals(oldFolder.getId());
        } else return oldEntity instanceof Footer && newEntity instanceof Footer;
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        Entity oldEntity = mOldList.get(oldItemPosition);
        Entity newEntity = mNewList.get(newItemPosition);
        if (oldEntity instanceof Header && newEntity instanceof Header) {
            Header newHeader = (Header) newEntity;
            Header oldHeader = (Header) oldEntity;
            return newHeader.getTitle().equals(oldHeader.getTitle());
        } else if (oldEntity instanceof File && newEntity instanceof File) {
            File newFile = (File) newEntity;
            File oldFile = (File) oldEntity;
            return newFile.getTitle().equals(oldFile.getTitle()) && newFile.getVersion() == oldFile.getVersion();
        } else if (oldEntity instanceof Folder && newEntity instanceof Folder) {
            Folder newFolder = (Folder) newEntity;
            Folder oldFolder = (Folder) oldEntity;
            return newFolder.getTitle().equals(oldFolder.getTitle()) && newFolder.getFilesCount() == oldFolder.getFilesCount();
        } else return oldEntity instanceof Footer && newEntity instanceof Footer;
    }
}

File: app_manager/appmanager/src/main/java/app/editors/manager/ui/adapters/diffutilscallback/EntityDiffUtilsCallback.kt
package app.editors.manager.ui.adapters.diffutilscallback

import app.documents.core.network.manager.models.base.Entity
import app.documents.core.network.manager.models.explorer.CloudFile
import app.documents.core.network.manager.models.explorer.CloudFolder
import app.editors.manager.mvp.models.list.Footer
import app.editors.manager.mvp.models.list.Header
import app.editors.manager.mvp.models.list.RecentViaLink
import app.editors.manager.mvp.models.list.Templates

class EntityDiffUtilsCallback(
    newList: List<Entity>,
    oldList: List<Entity>
) : BaseDiffUtilsCallback<Entity>(newList, oldList) {

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldEntity = mOldList[oldItemPosition]
        val newEntity = mNewList[newItemPosition]

        return when {
            oldEntity is Header && newEntity is Header -> true
            oldEntity is Templates && newEntity is Templates -> true
            oldEntity is RecentViaLink && newEntity is RecentViaLink -> true
            oldEntity is CloudFile && newEntity is CloudFile -> newEntity.id == oldEntity.id
            oldEntity is CloudFolder && newEntity is CloudFolder -> newEntity.id == oldEntity.id
            oldEntity is Footer && newEntity is Footer -> true
            else -> false
        }
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldEntity = mOldList[oldItemPosition]
        val newEntity = mNewList[newItemPosition]

        return when {
            oldEntity is Templates && newEntity is Templates -> true
            oldEntity is RecentViaLink && newEntity is RecentViaLink -> true
            oldEntity is Header && newEntity is Header -> newEntity.title == oldEntity.title

            oldEntity is CloudFile && newEntity is CloudFile ->
                newEntity.title == oldEntity.title
                        && newEntity.version == oldEntity.version
                        && newEntity.updated == oldEntity.updated
                        && newEntity.thumbnailUrl == oldEntity.thumbnailUrl
                        && newEntity.customFilterEnabled == oldEntity.customFilterEnabled
                        && newEntity.isEditing == oldEntity.isEditing
                        && newEntity.thumbnailStatus == oldEntity.thumbnailStatus
                        && newEntity.formFillingStatusType == oldEntity.formFillingStatusType

            oldEntity is CloudFolder && newEntity is CloudFolder ->
                newEntity.title == oldEntity.title
                        && newEntity.filesCount == oldEntity.filesCount
                        && newEntity.updated == oldEntity.updated

            oldEntity is Footer && newEntity is Footer -> true

            else -> false
        }
    }
}
